src/trio/_tests/test_subprocess.py

```diff
@@ -765,3 +765,58 @@ async def wait_and_tell(proc: Process) -> None:
             # for everything to notice
             await noticed_exit.wait()
         assert noticed_exit.is_set(), "child task wasn't woken after poll, DEADLOCK"
+
+
+@pytest.mark.skipif(not posix, reason="posix only")
+async def test_shells_killed_by_default() -> None:
+    async with _core.open_nursery() as nursery:
+        proc = await nursery.start(
+            partial(
+                trio.run_process,
+                'python -u -c "import os, time; print(os.getpid()); time.sleep(5)" | cat',
+                shell=True,
+                stdout=subprocess.PIPE,
+            )
+        )
+
+        # we need to wait until the process is started before cancelling it,
+        # otherwise, it never prints a PID (in fact the shell might not even
+        # have the chance to spawn children either)
+        await trio.sleep(0.1)
+        nursery.cancel_scope.cancel()
+
+    chunks = []
+    with trio.move_on_after(0.1):
+        async with proc.stdout:
+            async for c in proc.stdout:
+                chunks.append(c)  # noqa: PERF401
+
+    # give the child time to wind down execution (this is unfortunate)
+    await trio.sleep(0.1)
+
+    child_pid = int(b"".join(chunks))
+    with pytest.raises(OSError, match="No such process"):
+        os.kill(child_pid, 0)
+
+
+@pytest.mark.skipif(not posix, reason="posix only")
+async def test_process_group_SIGKILL_escalation(mock_clock: MockClock) -> None:
+    mock_clock.autojump_threshold = 0.1
```

Comment:
Unfortunately, all the tests for this need time for processes to *do* actual things. They would probably be better off with a higher sleep time and then be marked slow, but that kinda sucks?

Reply:
Turns out it's necessary for the Alpine and pypy-3.10 runners.

Reply:
And yet this somehow times out so I... don't know. This is so weird.